ingestion-sink/src/main/java/com/mozilla/telemetry/ingestion/sink/transform/PubsubMessageToTemplatedString.java

```diff
@@ -17,18 +26,77 @@
 
   public final String template;
 
-  public PubsubMessageToTemplatedString(String template) {
-    this.template = template;
+  public static PubsubMessageToTemplatedString of(String template) {
+    return new PubsubMessageToTemplatedString(template, DestinationType.OTHER);
+  }
+
+  public static PubsubMessageToTemplatedString forBigQuery(String template) {
+    return new PubsubMessageToTemplatedString(template, DestinationType.BIGQUERY);
   }
 
   @Override
   public String apply(PubsubMessage message) {
-    String batchKey = StringSubstitutor.replace(template,
-        DerivedAttributesMap.of(message.getAttributesMap()));
+    final String batchKey;
+    if (destinationType == DestinationType.BIGQUERY) {
+      batchKey = applyForBigQuery(message);
+    } else {
+      batchKey = applyForOther(message);
+    }
     if (batchKey.contains("$")) {
       throw new IllegalArgumentException("Element did not contain all the attributes needed to"
           + " fill out variables in the configured template: " + template);
     }
     return batchKey;
   }
+
+  private String applyForOther(PubsubMessage message) {
+    return StringSubstitutor.replace(template, DerivedAttributesMap.of(message.getAttributesMap()));
+  }
+
+  private String applyForBigQuery(PubsubMessage message) {
+    Map<String, String> attributes = new HashMap<>(message.getAttributesMap());
+
+    // We coerce all docType and namespace names to be snake_case and to remove invalid
+    // characters; these transformations MUST match with the transformations applied by the
+    // jsonschema-transpiler and mozilla-schema-generator when creating table schemas in BigQuery.
+    final String namespace = attributes.get(Attribute.DOCUMENT_NAMESPACE);
+    final String docType = attributes.get(Attribute.DOCUMENT_TYPE);
+    if (namespace != null) {
+      attributes.put(Attribute.DOCUMENT_NAMESPACE, getAndCacheNormalizedName(namespace));
+    }
+    if (docType != null) {
+      attributes.put(Attribute.DOCUMENT_TYPE, getAndCacheNormalizedName(docType));
+    }
+
+    attributes = Maps.transformValues(DerivedAttributesMap.of(attributes),
+        v -> v.replaceAll("-", "_"));
+
+    return StringSubstitutor.replace(template, attributes);
+  }
```

Comment:
This logic is almost exactly copied from KeyByBigQueryTableDestination. We will need a bigger refactor in order to support writing into live tables in ingestion-sink, and that would be the point at which to consolidate logic and remove this duplication.